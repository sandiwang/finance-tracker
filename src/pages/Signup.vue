<template>
  <div class="un-auth-page">
  	<v-layout row wrap fill-height light>
      <v-flex xs12 offset-md7 md5 d-flex light align-center class="main">
		    <v-layout row wrap justify-center>    
		    	<v-flex xs8>
		    		<v-form id="signup-form" v-model="formValid" @submit.prevent="userSignup">
		        	<v-layout wrap>
			        	<v-flex xs12 class="px-5 text-xs-center">
			        		<div class="headline mb-4">Sign Up</div>
			        		<v-layout wrap>
			        			<v-flex xs6 class="pr-1 text-xs-left">
			        				<v-text-field
							        	name="fname"
							        	type="text"
							        	label="First Name"
							        	id="fname"
							        	v-model="fname"
							        	:rules="validRules.name"
							        	required
							        ></v-text-field>
			        			</v-flex>
			        			<v-flex xs6 class="pl-1 text-xs-left">
			        				<v-text-field
							        	name="lname"
							        	type="text"
							        	label="Last Name"
							        	id="lname"
							        	v-model="lname"
							        	:rules="validRules.name"
							        	required
							        ></v-text-field>
			        			</v-flex>
			        		</v-layout>
			        	</v-flex>
			        	<v-flex xs12 class="text-xs-center px-5">
			          	<v-text-field
					          name="email"
					          type="text"
					          label="Email"
					          id="email"
					          v-model="email"
					          :rules="validRules.email"
					          required
					          class="text-xs-left"
					        ></v-text-field>
					        <v-text-field
					          name="password"
					          type="password"
					          label="Password"
					          id="password"
					          v-model="password"
					          :rules="validRules.password"
					          hint="Must contain at least one number and one uppercase and lowercase letter, and at least 8 or more characters"
					          persistent-hint
					          required
					          class="text-xs-left"
					        ></v-text-field>
					        <v-text-field
					        	name="confirmPassword"
					        	type="password"
					        	label="Confirm Password"
					        	id="confirmPassword"
					        	v-model="passwordConfirm"
					        	required
					        	class="text-xs-left mt-3"
					        	:rules="validRules.confirmPassword"
					        ></v-text-field>
					        <v-alert type="error" :value="alert" transition="slide-y-transition" class="text-xs-left mb-3 form-message">{{ getError }}</v-alert>
					        <v-btn type="submit" :disabled="!formValid" color="primary" class="mt-2">Sign Up</v-btn><br>
						    	<div class="links my-4">Or <router-link class="my-4" to="/signin">sign in</router-link></div>
			        	</v-flex>
			      	</v-layout>
			      </v-form>
			    </v-flex>
		    </v-layout>
      </v-flex>
    </v-layout>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
	name: 'Signup',
	data () {
		return {
			email: '',
			password: '',
			passwordConfirm: '',
			fname: '',
			lname: '',
			alert: false,
			formValid: false,
			validRules: {
				name: [
	        v => !!v || 'Name cannot be empty'
	      ],
				email: [
	        v => !!v || 'Email cannot be empty',
	        v => /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/.test(v) || 'Email not valid'
	      ],
	      password: [
	      	v => !!v || 'Password cannot be empty',
	      	v => /(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{8,}/.test(v) || 'Password not valid'
	      ],
	      confirmPassword: [
	      	v => (v === this.password) || 'Passwords don\'t match'
	      ]
			},
			defaultBanner: '/static/banner-default.jpg'
		}
	},
	computed: {
		...mapGetters([
			'getError'
		]),
		comparePasswords () {
			return this.password === this.passwordConfirm
		}
	},
	methods: {
		userSignup () {
			if( this.password !== this.passwordConfirm ) return;
			this.$store.dispatch('userSignup', {
				email: this.email,
				password: this.password,
				fname: this.fname,
				lname: this.lname,
				banner: this.defaultBanner,
			})
		}
	},
	watch: {
		getError (value) {
			if(value) this.alert = true
		},
		alert (value) {
			if(!value) this.$store.dispatch('setError', null)
		}
	}
}
</script>