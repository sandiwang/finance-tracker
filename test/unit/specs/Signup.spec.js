import { shallow, createLocalVue, mount } from 'vue-test-utils'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import Signup from '@/pages/Signup'

const localVue = createLocalVue()

localVue.use(Vuex)
localVue.use(Vuetify)

describe('Signup.vue', () => {
	let store
	let getters

	beforeEach(() => {
		getters = {
			getError: () => null
		}

    store = new Vuex.Store({
      getters
    })
  })

	it('renders the correct link to sign in page', () => {
		const wrapper = mount(Signup, { store, localVue })
    const link = wrapper.find('.links')

    expect(link.html()).toMatchSnapshot()
  })


})

/* ====================
	Example

	import Vue from 'vue'
	import HelloWorld from '@/pages/HelloWorld'

	describe('HelloWorld.vue', () => {
	  it('should render correct contents', () => {
	    const Constructor = Vue.extend(HelloWorld)
	    const vm = new Constructor().$mount()
	    expect(vm.$el.querySelector('.hello h1').textContent)
				.toEqual('Welcome to Your Vue.js App')
	  })
	})

======================= */