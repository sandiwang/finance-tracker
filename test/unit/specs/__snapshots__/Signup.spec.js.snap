// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup.vue renders the correct link to sign in page 1`] = `
<div class="links my-4">Or
    <router-link to="/signin" class="my-4">sign in</router-link>
</div>
`;
